File: PyProject/leetcode/binary-tree-level-order-traversal.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def helper(nodes: List[TreeNode], result: List[List[int]]):
    next_nodes = []
    cur_vals = []
    for node in nodes:
        if node is None:
            continue
        cur_vals.append(node.val)
        next_nodes.append(node.left)
        next_nodes.append(node.right)
    if cur_vals:
        result.append(cur_vals)
        helper(next_nodes, result)


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        result = []
        nodes = [root]
        helper(nodes, result)
        return result

File: GoProject/bytedance/queue/binary-tree-level-order-traversal.go
package queue

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type levelOrderData struct {
	Data [][]int
}

func levelOrderHelper(nodes []*TreeNode, result *levelOrderData) {
	if len(nodes) == 0 {
		return
	}

	newNodes := []*TreeNode{}
	newResults := []int{}

	for _, node := range nodes {
		if node == nil {
			continue
		}
		newResults = append(newResults, node.Val)
		newNodes = append(newNodes, node.Left)
		newNodes = append(newNodes, node.Right)
	}

	if len(newResults) != 0 {
		result.Data = append(result.Data, newResults)
	}
	levelOrderHelper(newNodes, result)
}

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	result := &levelOrderData{Data: [][]int{}}

	levelOrderHelper([]*TreeNode{root}, result)
	return result.Data
}
